cpp/src/io/parquet/reader_impl.cpp

```diff
@@ -120,8 +147,14 @@ void reader::impl::decode_page_data(size_t skip_rows, size_t num_rows)
   chunks.host_to_device(_stream);
   chunk_nested_valids.host_to_device(_stream);
   chunk_nested_data.host_to_device(_stream);
+  chunk_nested_str_data.host_to_device(_stream);
 
+  // TODO: explore launching these concurrently with a stream pool
```

Comment:
This was an offline suggestion from @nvdbaranec for #12948, where I was once doing just what you suggest in `gpuDecodePageData`. If I am remembering and understanding correctly, branching to call the different decoders increases the shared memory requirements for all, even if the other kernels aren't called.  This approach was seen as a stop-gap solution.  It would ultimately be nice to split the pages up into batches, with each batch being sent to the proper decoder kernel. Down the road there will be additional kernels for the delta encodings.

Reply:
Another thing to consider here:  by calling these on the same stream we are losing a decent amount of parallelism.  These kernels won't execute at the same time, so it's possible to get into a situation like:

- GPU has 100 SMs
- 50 pages contain string data
- 50 pages contain int data

At any given time, we will only be using half the SMs.   I think it would be worth (as a followup) to either figure out how to conglomerate shared memory usage so that these can be run in one invocation, or to investigate using multiple streams.   

I'm doing something like with the an upcoming PR.  Fundamentally:

This:
```
__global__ kernel()
{
    if(a){
       path_a();
   } else {
      path_b();
   }
}

void path_a()
{
    __shared__ x;
}

void path_b()
{
    __shared__ y;
}
```

Becomes:

```
struct shared_stuff{
  ...
};

__global__ kernel()
{
  __shared__ shared_stuff s;
   if(a){
      path_a(s);
   } else {
      path_b(s);
   }
}

void path_a(shared_stuff s)
{
}

void path_b(shared_stuff s)
{
}
```

Since the shared stuff for a and b is pretty much the same (buffers for rep/def levels, some cub scratch space, etc) it ends up removing most duplication.